add Calendar

igarashi-naoki opened the pull request:
カレンダープログラムの追加

igarashi-naoki committed:
add Calendar

igarashi-naoki committed:
Add new line after last day.    Add --help option.

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,69 @@
+  print "\e[7m" if TODAY.eql?(Date.new(year, month, day))
+  print format('%<x>2d', x: day)
+  print "\e[0m  "
+  print "\n" if weekday % 7 == 6
```
土曜日の判定について、 `day` には `Date` のオブジェクトが入っているはずなので、以下メソッドを活用できないでしょうか？
https://docs.ruby-lang.org/ja/latest/method/Date/i/saturday=3f.html

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,69 @@
+(1..Date.new(year, month, -1).day).each do |day|
+  print "\e[7m" if TODAY.eql?(Date.new(year, month, day))
+  print format('%<x>2d', x: day)
+  print "\e[0m  "
```
こちらの制御コードは、今日以外の日付でも出力されてしまいそうですが問題ないでしょうか？

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,69 @@
+  exit
+end
+
+if !options[:month].nil?
```
少し判定が冗長なように見えました。オプションがある時は [こちらのブロック](https://github.com/igarashi-naoki/ruby-practices/pull/1/files#diff-b8defe7dbf15f532417d0404562fc1b2121c2f9c42360b2a4c4e843905b418d8R12) が呼ばれ、オプションが呼ばれないという挙動となっているので、現状はオプションの有無を2箇所で判定している状態になっていると言えます。

ブロック内を上手く活用すれば、このif文はいらなくなるはずです。

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,69 @@
+    begin
+      puts "'#{options[:month]}' is neither a month number (1..12) nor a name"
+      exit
+    rescue SystemExit
```
この例外はどういったケースで起こりうるものでしょうか？

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,69 @@
+weekday = Date.new(year, month, 1).wday
+(0...weekday).each { print '　　' }
+
+TODAY = Date.today
```
[こちら](https://github.com/igarashi-naoki/ruby-practices/pull/1/files#diff-b8defe7dbf15f532417d0404562fc1b2121c2f9c42360b2a4c4e843905b418d8R6) でも `Date.today` を使っているのでまとめられると良いですね。

igarashi-naoki committed:
calendar出力周りのリファクタリング

igarashi-naoki reviewed (commented):
再度レビューお願いします🙇‍♀️

### 変更点
1. 定数`TODAY` を再利用するように変更

2. 月初日のスペース出力を変更

3. `Range`オブジェクトで`Date`オブジェクトを渡すように変更

4. 改行判定時に`saturday?`メソッドを使用

### 変更理由

1. デフォルト年月として、今日の年月を用いており、定数`TODAY`の再利用が可能だったため。

2. ここの２行でしか使用されない変数`weekday`が不要。またここでのループ処理の意図としては`times`が適切。

3. 頂いたレビューからブロック内の処理を考慮すると、`Date`オブジェクトを渡すことが適切だと判断。以前は整数値を渡していたが、背景色反転や土曜日改行の判定を考え修正。

4. レビューをもとに修正。直感的でわかりやすいメソッドを使うべきだった。(知っておくべきだった)

igarashi-naoki committed:
コマンドライン引数及び例外処理に関する修正

igarashi-naoki committed:
-mオプション指定時に必ず引数を持つように変更。またそれに付随する例外処理の追加

以前は-mオプション指定時に引数の不足を許容していた。（デフォルト値は当月）
そのため、OptionParserの各オプション処理のブロック内でmの値がnilになることがあり、mのnil判定を行っていた。
しかし、Linuxのcalコマンドでは-mの引数不足でエラーとなるため、それに準拠するように変更。

igarashi-naoki commented:
### 変更点

1. オプション受け取りの処理を`OptionParser`のブロック内で完結させるように変更

2. また、１のブロック内の処理をそれぞれメソッドとして抜き出した

3. `-m`オプション指定時に必ず引数を取るように変更

4. `SystemExit`に関係する不要な例外処理を行わないように変更



### 変更理由

1.  以前はハッシュへの格納と例外判定を別の箇所で行っていたが、それを一か所にまとめるため。

2. メインプログラムの可読性向上のため。

3. [mオプション指定時に必ず引数を持つように変更](https://github.com/igarashi-naoki/ruby-practices/pull/1/commits/26c3d9e59fe0da4d2359d69176d286e6acec389a)

4. レビューの内容をもとに修正

igarashi-naoki committed:
-mオプションの引数に関してnil判定を行わないように変更

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,67 @@
+
+def get_arg_year(arg_year)
+  # 文字列を数値変換できない場合ArgumentError
+  if (1..9999).cover?(Integer(arg_year))
```
ここですが、 `Integer(arg_year)` ではなく `arg_year.to_i` を使うようにすれば `nil` は `0` に変換されるので、メソッドの呼び元で行っている `unless y.nil?` を省略できますね。

igarashi-naoki replied:
-yオプションも引数を必ず取るように修正しました。

yuuu commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,67 @@
+
+(Date.new(year, month, 1)..Date.new(year, month, -1)).each do |day|
+  print "\e[7m" if TODAY.eql?(day)
+  print format("%<x>2d\e[0m  ", x: day.day)
```
このコードでも今日以外の日付で制御コードが出力されそうですが、問題ないと判断されたということでしょうか？
https://github.com/igarashi-naoki/ruby-practices/pull/1#discussion_r1225937831

igarashi-naoki replied:
レビューありがとうございます。

`Date`クラスに問題がなかったたため、問題ないと判断していましたが、`eql?`メソッドでハッシュ値の衝突が起こる可能性を考慮できていませんでした。🙇‍♀️  
また、`Date`オブジェクトの内容が同じでも必ず同じ `object_id `になるわけでもないと分かったので、`equal?`メソッドも不適切だと理解できました。

`Date`クラスのインスタンスメソッドである[`===`](https://docs.ruby-lang.org/ja/latest/class/Date.html#I_--3D--3D--3D)を使用したいと思います。

igarashi-naoki replied:
~~`eql?`メソッドでハッシュ値の衝突が起こる可能性を考慮できていませんでした。🙇‍♀️~~　
↓
`Date`クラスで [`eql?`](https://docs.ruby-lang.org/ja/latest/class/Object.html#I_EQL--3F)メソッドはオーバーライドされていないので、[`equal?`](https://docs.ruby-lang.org/ja/latest/class/Object.html#I_EQUAL--3F)メソッドと同じ挙動になるので不適切。

yuuu replied:
指摘と回答が少しズレているようです。
自分がコメントしているのは「今日以外の日付であっても `\e[0m` を出力する方針としているようだが問題ないか？」という点でした。修正後のコードでもこの点については変わっていない認識ですし、「指定をリセットする」というものなので問題ない、という判断もありだと思っています。

回答いただいているのは「今日の判定方法」に関するものでそちらは `===` を使う方針で良いと思います。

igarashi-naoki replied:
コメントありがとうございます。🙇‍♀️　

ここでは指定をリセットする制御コードなので問題ない、という判断をしました。

「（今日の判定方法に問題があるので）今日以外の日付で制御コード`"\e[7m"`が出力される可能性がある。」と勘違いしておりました、お手数おかけしました。😖

yuuu replied:
なるほど、日本語は難しいですね 😅 私の方も言葉足らずですみませんでした 🙏

igarashi-naoki committed:
当日判定処理の変更、オプション処理の変更

背景色反転のための当日判定処理において、当日以外も背景色が反転してしまう可能性があったため修正。

オプション処理において-yオプションも引数を必ず取るように修正。
またオプション引数を数値変換する際に、意図しない挙動が起こる可能性があったため修正。

例:
Integerクラスによる変換は文字列の先頭が0の場合、8進数として解釈される。
Integer("0377")  #=> 255

to_iメソッドによる変換は数値＋文字列の場合、数値のみを抽出して変換する。
"2023aa".to_i  #=> 2023
